### ISSUE TITLE
h265 4K video headset frame rate 

### ISSUE THREAD

**Issue by robertncoomber** [RenderHeads/UnityPlugin-AVProVideo#426]
**Describe the bug**
The Frame Rate of the Oculus Rift S Headset is heavily reduced when playing my hevc 4k video. Upon starting the video the frame rate drops to between 45 and 60 fps, and rises back to 80 fps when paused. This happens on the 3840x3840 video but the frame rate stays at 80 fps when playing a 2048 * 2048 version of the same footage. The PC should be more than capable of decoding the 4K hevc and smooth playback. The frame rate during playback in the editor without VR enabled seems just fine, which is interesting. 

**Your Setup (please complete the following information):**
 - Unity version: 2019.4.3f1
 - AVPro Video version: Version 1.11.3 
 - Operating system version: Windows 10 Pro 1909
 - Device model: 
     - i7 3770K 
     - 16 gb ram
     - GTX 1070
     - 1 tb ssd
     - hevc codec installed

- Video specs (resolution, frame-rate, codec, file size): 3840*3840, 30 fps, h265, 45 mb, 12 seconds 

 > ffmpeg -i input.mov -vf scale=3840x3840 -r 30 -c:v libx265 -preset fast -crf 18 -maxrate 60m -pix_fmt yuv420p -movflags faststart output.mp4

[Demo Videos](https://drive.google.com/drive/folders/1hJnCHUiP6CbZKtrbOy85F96ig_mSAB2f?usp=sharing)

**To Reproduce**
1. Open new unity project
2. Import AVpro
3. Go to player settings -> XR Settings
4. Check Virtual Reality Supported
5. Drag and drop 360SphereVideo prefab into scene
6. Point Media Player path to debugger.mp4 (listed in drive link), in either streamingassets or absolute path
7.  Play Video
8. Open C:\Program Files\Oculus\Support\oculus-diagnostics\OculusDebugTool.exe if you would like to view profiler

**Logs**
[Oculus Profiler](https://drive.google.com/file/d/1lH4083mKN0uyecDTcZGJ1B3crzj6xEDn/view?usp=sharing)
[Oculus Lost Frame Capture Logs](https://drive.google.com/file/d/1D27AIIYvDgkciKM8X3B1wKblAVqwE2id/view?usp=sharing)

**Videos**
The video shows the in game frame rate monitor while playing and pausing the video. 
[Screen Capture Video](https://drive.google.com/file/d/18fMoyuzWzSUODb4WBc4HnLD5jukD7sjk/view?usp=sharing)

**Comment by robertncoomber** [comment]
Any word on this issue? I was doing some more testing and I noticed that the resized videos were accidentally converted to h.264. All the video files in the google drive are each h.265 now. During some researching and testing, I found a couple video players on the Oculus store that use unity + AVpro which played the video with no fps issues. PlayaVR is one in specific I know which uses AV Pro and Unity which can play these video just fine. I can't figure out why or how it is playing just fine with their app however however with mine, there are these issues. In addition, I did a standalone quest build with AVpro android playing the video and the frame rate was steady at 72 fps. It seems strange the quest can play the video just fine but my PC cant. I am wondering if there was an update to AVpro at some point which changed the way the hevc videos are decoded? Let me know if you need any other information. Any help is appreciated. Thanks.

**Comment by AndrewRH** [comment]
Hi @robertncoomber  have you tried using the Unity profiler to see where the frame time is going?

Thanks,

**Comment by robertncoomber** [comment]
Hi @andrewRH,

As I was profiling, I decided to try unity 2018 rather than 2019 or 2020. This framerate issue that I am talking about is not affecting unity 2018 with the most current Oculus integration plugin. For testing, I am using 2018.4.22 and 2019.2.9. 

While I was profiling, I found some strange results. One observation I noticed is that the Unity Frame rate within the stat tab in the game window does not seem to always correlate with Oculus' Debug Tool **ODT** fps recorder.  The Unity Stats fps for both 2018 and 2019 versions of Unity showed similar frame rates while playing the video ≈90 fps which isn't always correct. 

While playing the video on Unity 2018, the **ODT** frame rate stays constantly at 80 fps. The same video on unity 2019 hovers between 45 and 60 on **ODT** fps.  

I think it might be related to Unity's XR plugin management system in 2019 and up. When reviewing the profiler data in 2019, that XR.WaitForGPU takes 7+ms but when I got check the GPU, the GPU time was about 1ms which means there is another issue at play. 
![VdNKsrOpji](https://user-images.githubusercontent.com/19243496/88877405-3d50c280-d1da-11ea-970d-9c6eca2f02ba.png)
![rLTl7ci5DQ](https://user-images.githubusercontent.com/19243496/88877412-3f1a8600-d1da-11ea-84c5-46c81aac7ef2.png)
[2018and2019profilerdata.zip](https://github.com/RenderHeads/UnityPlugin-AVProVideo/files/4998616/2018and2019profilerdata.zip) Here is the unity profiler .data for both 2018 and 2019 with the video playing if you might need them.


[This thread here](https://forum.unity.com/threads/earlyupdate-xrupdate-taking-about-75-of-main-thread-cpu-time-is-this-normal.920234/) seems to possibly be related. However this is using a later version of unity. I plan to try on this never version of unity within the next couple of days. 

It is also very strange that this issue seems to be only effecting high resolution 4k video playback in this instance. 

I am reverting my project down to 2018 for now however, I would like to figure this out so I can access some of the 2019 features in the future. If I find out any more information I will post it here.
